### PR TITLE
Divide eav attribute cache

### LINKED ISSUE #10978: Multi website - customer create form - Attribute Cache not taking account the current website/store
### Preconditions
1. Magento CE 2.1.9
2. Multi website / and multi stores (B2B (with languages : FR, EN) + B2C (with languages FR, EN, JP)
3. Php 5.6

### Steps to reproduce

1. For the B2b website I want to display Tax / Vat field. And I don't want to display DOB field
2. For the B2C website I don't want to display Tax / Vat field, but I do want to display DOB field (optional).
3. So I set up these parameters in the backend Stores => Settings/Configuration => Customers / Customers configuration => Name and Address Options

| Website | Setting | Value |
| - | - | - |
| B2B | Show Date of birth | No
|B2B | Show Tax/VAT Number | Optional |
| B2C | Show Date of birth | Otptional |
| B2C | Show Tax/VAT Number | No |

### Expected result
1. I flush all cache
2. I go to B2B website on customer/account/create => generate form and display Tax / Vat field as optional and hide Dob field
3. I go to B2C website on customer/account/create => generate form and display Dob field as optional but hide Tax / Vat field.

### Actual result

1. I flush all cache
2. I go to B2B website on customer/account/create => generate form and it display Tax / Vat field as optional and hide Dob field => OK!
3. I go to B2C website on customer/account/create => generate form and it display the same form as B2B website => NOK!

4. I flush cache
5. I go to B2C website on customer/account/create => generate form and it display Dob field as optional but hide Tax / Vat field => OK!
6. I go to B2B website on customer/account/create => gernerate form and it display the same form as B2C website => NOK!

going deeper in the code I think the problem is related to the AttributeCache class :

Magento\Customer\Block\Widget\Taxvat
Magento\Customer\Block\Widget\AbstractWidget::_getAttribute (:96)
Magento\Customer\Model\Metadata::getAttributeMetadata (:62)
Magento\Eav\Model\Entity\AttributeCache::saveAttributes (:108)
=> The cache Key does not take the website_id or store_id into account.
So once it is cached on a website it will give the same result on the other websites.

### PR DESCRIPTION
**Pull request by vshcheglov**
### Description (*)
We have an ability to choose different attribute settings per website in Magento admin, but have no eav attribute cache separation. So our attribute settings will be depended on which website will be opened first after eav cache cleaning.

**This is a duplicate of https://github.com/magento/magento2/pull/33061
but have additional changes and based on 2.4-develop branch and being merged into it.**

### Fixed Issues (if relevant)

1. Fixes magento/magento2#10978

### Manual testing scenarios (*)
1. Open clean Magento instance

Creating new website and store
1. Go to Stores -> Settings -> All Stores
1. Select Create Website
1. Create a new website
1. Select Create Store
1. Create a new store and assign it to a newly created website
1. Select Create Store View
1. Create a new enabled store view and assign it to newly created store
1. Remember store view code
1. Go to System -> Cache Management
1. Flush Cache Storage

Filling required settings
1. Go to Stores -> Configuration -> Customers -> Customer Configuration -> Name and Address Options
1. Select Main Website scope
1. Set Date of Birth to No
1. Set Show Tax/VAT Number to Optional
1. Save config
1. Select new website scope
1. Set Date of Birth to Optional
1. Set Show Tax/VAT Number to No
1. Save config

Enabling store code in urls for fast website switching
1. Select Default Config scope
1. Go to General -> Web -> Url Options
1. Change Add Store Code to Urls to Yes
1. Save config

Reproducing the bug
1. Go to System -> Cache Management
1. Flush cache storage
1. Important! First of all, go to Main Website Storefront Create Account Page
1. Go to https://yourmagento.com/default/customer/account/create/
1. We see Tax/Vat Number field and don't see Dob field
1. Then go to new created website Create Account Page
1. Go to https://yourmagento.com/new-store-view-code/customer/account/create/

Expected:
1. We don't see Tax/Vat Number field and see Dob field

Actual:
1. We see Tax/Vat Number field and don't see Dob field

Then another case:
1. Flush cache storage
1. Important! First of all, go to New Website Storefront Create Account Page
1. Go to yourmagento.com/new-store-view-code/customer/account/create/
1. We don't see Tax/Vat Number field and see Dob field (as expected sooner above)
1. Now go to Main Website Storefront create account page
1. Go to https://yourmagento.com/default/customer/account/create/

Expected:
1. We see Tax/Vat Number field and don't see Dob field

Actual:
1. We don't see Tax/Vat Number field and see Dob field

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Eav/Model/Config.php [hunk 1]
```diff
@@ -174,6 +182,7 @@ public function __construct(
         \Magento\Framework\Validator\UniversalFactory $universalFactory,
         SerializerInterface $serializer = null,
         ScopeConfigInterface $scopeConfig = null,
+        StoreManagerInterface $storeManager = null,
```
Comment (ihor-sviziev) [inline app/code/Magento/Eav/Model/Config.php]: Could you please move it to the last position?

File: app/code/Magento/Eav/Model/Config.php [hunk 2]
```diff
@@ -461,16 +471,16 @@ public function getEntityType($code)
      * @param string $entityType
      * @return $this
      */
-    protected function _initAttributes($entityType)
+    protected function _initAttributes($entityType, $storeId)
```
Comment (ihor-sviziev) [inline app/code/Magento/Eav/Model/Config.php]: I think we should add some default values, like "global", to make this change backward compatible.
```suggestion
    protected function _initAttributes($entityType, $storeId = 0)
```

Or make it nullable and try to retrieve the current store in such case.
```suggestion
    protected function _initAttributes($entityType, $storeId = null)
```
What do you think?

File: app/code/Magento/Eav/Model/Config.php [hunk 3]
```diff
@@ -482,19 +492,19 @@ protected function _initAttributes($entityType)
...
             $this->saveAttribute($attributeObject, $entityTypeCode, $attributeObject->getAttributeCode());
-            $this->_attributeData[$entityTypeCode][$attribute['attribute_code']] = $attributeObject->toArray();
+            $this->_attributeData[$storeId][$entityTypeCode][$attribute['attribute_code']] = $attributeObject->toArray();
         }
         if ($this->isCacheEnabled()) {
             $this->_cache->save(
-                $this->serializer->serialize($this->_attributeData[$entityTypeCode]),
-                self::ATTRIBUTES_CACHE_ID . $entityTypeCode,
+                $this->serializer->serialize($this->_attributeData[$storeId][$entityTypeCode]),
+                self::ATTRIBUTES_CACHE_ID . '-' . $this->getStoreId() . '-' . $entityTypeCode,
```
Comment (ihor-sviziev) [inline app/code/Magento/Eav/Model/Config.php]: I think we should use $storeId variable instead
```suggestion
                self::ATTRIBUTES_CACHE_ID . '-' . $storeId . '-' . $entityTypeCode,
```

File: app/code/Magento/Eav/Model/Config.php [hunk 4]
```diff
@@ -583,7 +593,7 @@ private function initSystemAttributes($entityType, $systemAttributes)
             return;
         }
 
-        $cacheKey = self::ATTRIBUTES_CACHE_ID . '-' . $entityTypeCode . '-preload';
+        $cacheKey = self::ATTRIBUTES_CACHE_ID . '-' . $this->getStoreId() . '-' . $entityTypeCode . '-preload';
```
Comment (ihor-sviziev) [inline app/code/Magento/Eav/Model/Config.php:598]: Can the system attributes be separated by the store? If not - I think we can use a "global" value

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Eav/Model/Config.php (hunk 1 above)
Comment (vshcheglov) [reply app/code/Magento/Eav/Model/Config.php]: Oh sure, also optional parameter

File: app/code/Magento/Eav/Model/Config.php (hunk 3 above)
Comment (vshcheglov) [reply app/code/Magento/Eav/Model/Config.php]: yep, and in initAttributesFromCache method too

File: app/code/Magento/Eav/Model/Config.php (hunk 2 above)
Comment (vshcheglov) [reply app/code/Magento/Eav/Model/Config.php]: I am absolutely sure that we need default value for the new argument but not sure which variant is better.
Let's retrieve the current store.

File: app/code/Magento/Eav/Model/Config.php (hunk 4 above)
Comment (vshcheglov) [reply app/code/Magento/Eav/Model/Config.php:598]: Method has the name initSystemAttributes but not only system attributes being loaded by it.
It loads attributes provided in attributesForPreload constructor argument which populated from di configuration.
So if we not split here we will get issue described in the pull request again.

But good news that we can remove splitting from cacheUserDefinedAttribute method.

Comment (vshcheglov) [reply app/code/Magento/Eav/Model/Config.php:598]: I renamed private initSystemAttributes method to initPreloadAttributes.
So for no one else gets confused anymore.

File: app/code/Magento/Eav/Model/Config.php [hunk 5]
```diff
@@ -183,6 +192,7 @@ public function __construct(
         $this->_universalFactory = $universalFactory;
         $this->serializer = $serializer ?: ObjectManager::getInstance()->get(SerializerInterface::class);
         $this->scopeConfig = $scopeConfig ?: ObjectManager::getInstance()->get(ScopeConfigInterface::class);
+        $this->storeManager = $storeManager ?: ObjectManager::getInstance()->get(StoreManagerInterface::class);
         $this->attributesForPreload = $attributesForPreload;
```
Comment (ihor-sviziev) [inline app/code/Magento/Eav/Model/Config.php]: Let's move it to the last position, the same as an argument in constructor
```suggestion
        $this->attributesForPreload = $attributesForPreload;
        $this->storeManager = $storeManager ?: ObjectManager::getInstance()->get(StoreManagerInterface::class);
```

File: app/code/Magento/Eav/Model/Config.php [hunk 6]
```diff
@@ -12,7 +12,9 @@
 use Magento\Framework\App\ObjectManager;
 use Magento\Framework\Exception\LocalizedException;
 use Magento\Framework\Model\AbstractModel;
+
```
Comment (Coosos) [inline app/code/Magento/Eav/Model/Config.php]: Remove empty line :)

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by vshcheglov** [comment]
@magento run all tests

**Comment by vshcheglov** [comment]
@magento run Static Tests

**Comment by vshcheglov** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
Hi @vshcheglov,

Thank you for your contribution!

I am planning to work on the failing tests, but before proceeding, I have tried to reproduce this issue on 2.4-develop branch and it is not reproducible.

To reproduce it, I am doing below steps.

1. Installed fresh 2.4-develop instance
2. Create a new Website, Store and Store View
image

3. Go to **Stores->Configuration->Customers->Customer Configuration->Name and Address Options**
4. Switch the scope **Main Website** and set **Show DOB** field to **No** and **Show TAX/VAT Number** to **Optional** and save config
Screenshot 2022-02-14 at 4 50 55 PM

5. Now switch the scope to **Second Website** and set **Show DOB** field to **Optional** and **Show TAX/VAT Number** to **No** and Save config
Screenshot 2022-02-14 at 4 54 51 PM

6. Enable Store code in URL's, switch to Default scope and go to **Stores->Confoguration->General->Web->Url Options->Add Store Code to Urls** and set to **Yes** and then Save Config
image

7. From **System->Cache Management** Flush Cache Storage

**Test Case:**
1. Go to Main Website Storefront Create Account Page
2. Go to https://yourmagento.com/default/customer/account/create/
3. We **see Tax/Vat Number** field and **don't see DOB** field
Screenshot 2022-02-14 at 5 02 45 PM

4. Then go to new created website Create Account Page
5. Go to https://yourmagento.com/new/customer/account/create/
6. We **see DOB** field and **don't see Tax/Vat Number** field
Screenshot 2022-02-14 at 5 05 06 PM

As we can see in both the above screenshots **the fields are rendered as expected!**

Can you please provide the detail steps to reproduce this issue if I am missing anything. In order to proceed further to fix the failed tests it will be really helpful.

Thank you in advance!

**Comment by vshcheglov** [comment]
@engcom-Lima I think the issue was resolved by this commit
https://github.com/magento/magento2/commit/e7386e9579adfcd26c64be06070d0cc70b8cebd8

**Comment by engcom-Lima** [comment]
Thank you for confirming @vshcheglov

Closing this PR as it has already been fixed and not reproducible on 2.4-develop.

Thanks!

**Comment by amenk** [comment]
@vshcheglov I applied that commit to a Magento 2.3.5 - and it seems to have a problem.

I am reproducing with the web API and a field which required in some websites.

I push to 

1. website A with required attribute
2. website B without required attribute in config (-> I should not get an error) Without the patch I get an error
3. website A again
4. website B -> now even with the patch I get an error

IMHO it is still caching, but only on the second attempt, because of this line:

https://github.com/magento/magento2/commit/e7386e9579adfcd26c64be06070d0cc70b8cebd8#diff-5fdab054d3795ae71d13da000e9a02aeabbfc2fb0b2f101595df14a07e932b29R810

**Pull request closed without merging**